### ISSUE TITLE
Unnecessary warning about undefined inline for constexpr function template when used in an unevaluated context

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#16992]
|  |  |
| --- | --- |
| Bugzilla Link | [16618](https://llvm.org/bz16618) |
| Version | trunk |
| OS | All |
| CC | @Axel-Naumann,@DougGregor,@zygoloid |

## Extended Description 
uneval.cpp:2:15: warning: inline function 'func<int>' is not defined [-Wundefined-inline]
constexpr int func();
              ^
uneval.cpp:5:17: note: used here
  return sizeof(func<int>());
                ^
1 warning generated.
```
blaikie@blaikie:~/dev/scratch$ cat uneval.cpp
template<typename T>
constexpr int func();

int main() {
  return sizeof(func<int>());
}
```

I'm not sure if this is really important/valid, but it seemed strange enough to file.

**Comment by serge-sans-paille** [comment]
No longer fails in main branch : https://godbolt.org/z/3Khbz986K
